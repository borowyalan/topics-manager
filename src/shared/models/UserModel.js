export class UserModel {
    email;
    name;
    nickname;


    static fromBackend(userData) {
        return Object.assign(new UserModel(), {
            email: userData.email,
            name: userData.name,
            nickname: userData.nickname
        });
    }
}
