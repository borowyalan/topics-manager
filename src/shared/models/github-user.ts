import { User } from './user';

export interface GithubUser extends User {
    idTokenPayload: string;
}
