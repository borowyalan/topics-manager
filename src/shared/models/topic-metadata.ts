export interface TopicMetadata {
    title: string;
    description: string;
}
