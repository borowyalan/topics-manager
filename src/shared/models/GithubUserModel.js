import { UserModel } from './UserModel';
export class GithubUserModel extends UserModel {
    static fromOAuth0(idTokenPayload) {
        return Object.assign(new GithubUserModel(), idTokenPayload);
    }
}
