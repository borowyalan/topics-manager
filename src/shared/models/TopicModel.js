import { UserModel } from './UserModel';
import { find } from 'lodash';

export default class TopicModel {
    title;
    description;
    likes;
    trainers;
    members;
    author; //GithubUserModel
    id;

    static fromBackendData(backendData) {
        const topic = Object.assign(new TopicModel(), backendData);
        topic.members = backendData.members
            ? backendData.members.map(member =>
                Object.assign(new UserModel(), member)
            )
            : [];
        topic.trainers = backendData.trainers
            ? backendData.trainers.map(trainer =>
                Object.assign(new UserModel(), trainer)
            )
            : [];
        return topic;
    }

    amIAttending(myself) {
        return myself && find(this.members, { email: myself.email });
    }

    amITrainer(myself) {
        return (
            myself &&
            (this.author.email === myself.email ||
                find(this.trainers, { email: myself.email }))
        );
    }
}
