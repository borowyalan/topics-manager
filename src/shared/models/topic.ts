import { find } from 'lodash';
import { TopicMetadata } from './topic-metadata';
import { User } from './user';

export interface Topic extends TopicMetadata {
    trainers: User[];
    members: User[];
    author: User;
    id: string;
}

export const isAttending = (topic: Topic, user: User) => {
    return Boolean(user && find(topic.members, { email: user.email }));
};

export const isTrainer = (topic: Topic, user: User) => {
    return Boolean(
        user &&
            (topic.author.email === user.email ||
                find(topic.trainers, { email: user.email }))
    );
};
