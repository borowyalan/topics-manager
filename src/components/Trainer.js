import React from 'react';
import PropTypes from 'prop-types';
import {UserModel} from '../shared/models/UserModel';

const Trainer = ({trainer, important}) => (
    <div style={important ? {backgroundColor: 'yellow'} : {}}>
        <p>{trainer.name}</p>
    </div>
);

Trainer.propTypes = {
    trainer: PropTypes.instanceOf(UserModel).isRequired,
    important: PropTypes.bool
};

export default Trainer;
