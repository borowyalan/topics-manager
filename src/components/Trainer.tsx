import React, { SFC } from 'react';
import styled from 'styled-components';
import { User } from '../shared/models/user';
import Text, { TextType } from './Text';

export interface TrainerProps {
    trainer: User,
    important?: boolean,
}

const Trainer: SFC<TrainerProps> = ({ trainer, important = false }) => (
    <StyledTrainer>
        <Text
            type={TextType.Basic}
            style={{ display: 'inline', textDecoration: (important ? 'underline' : 'none') }}>
            {trainer.name}
        </Text>
    </StyledTrainer>
);

const StyledTrainer = styled.li`
    margin: 15px 0;
`;

export default Trainer;
