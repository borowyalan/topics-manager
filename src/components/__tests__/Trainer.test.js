import React from 'react';
import {shallow} from 'enzyme';
import 'jest-styled-components';
import Adapter from 'enzyme-adapter-react-16';
import {configure} from 'enzyme/build/index';
import Trainer from '../Trainer';
import {UserModel} from '../../shared/models/UserModel';

configure({adapter: new Adapter()});

describe('<Trainer> component', () => {
    const someTrainer = UserModel.fromBackend({
        email: 'john@doe.com',
        name: 'John Doe'
    });

    let wrapper = shallow(<Trainer trainer={someTrainer}/>);

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

});
