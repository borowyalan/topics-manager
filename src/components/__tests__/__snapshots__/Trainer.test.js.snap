// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trainer> component should render 1`] = `
<div
  style={Object {}}
>
  <p>
    John Doe
  </p>
</div>
`;
